src/LanguageServer/Microsoft.CodeAnalysis.LanguageServer/HostWorkspace/FileBasedProgramsProjectSystem.cs

```diff
@@ -0,0 +1,143 @@
+    {
+        var documentFilePath = uri.ParsedUri is not null ? ProtocolConversions.GetDocumentFilePathFromUri(uri.ParsedUri) : uri.UriString;
+
+        // https://github.com/dotnet/roslyn/issues/78421: MAS should be its own workspace
+        if (_metadataAsSourceFileService.TryAddDocumentToWorkspace(documentFilePath, documentText.Container, out var documentId))
+        {
+            var metadataWorkspace = _metadataAsSourceFileService.TryGetWorkspace();
+            Contract.ThrowIfNull(metadataWorkspace);
+            var document = metadataWorkspace.CurrentSolution.GetRequiredDocument(documentId);
+            return document;
+        }
+
+        var primordialDoc = AddPrimordialDocument(uri, documentText, languageId);
+
+        if (uri.ParsedUri?.IsFile is true
+            && primordialDoc.Project.Language == LanguageNames.CSharp
+            && GlobalOptionService.GetOption(LanguageServerProjectSystemOptionsStorage.EnableFileBasedPrograms))
+        {
+            Contract.ThrowIfNull(primordialDoc.FilePath);
+            await BeginLoadingProjectAsync(primordialDoc.FilePath, primordialProjectId: primordialDoc.Project.Id);
+        }
+
+        return primordialDoc;
+
+        TextDocument AddPrimordialDocument(DocumentUri uri, SourceText documentText, string languageId)
+        {
+            var languageInfoProvider = _lspServices.GetRequiredService<ILanguageInfoProvider>();
+            if (!languageInfoProvider.TryGetLanguageInformation(uri, languageId, out var languageInformation))
+            {
+                Contract.Fail($"Could not find language information for {uri} with absolute path {documentFilePath}");
+            }
+
+            var workspace = Workspace;
+            var sourceTextLoader = new SourceTextLoader(documentText, documentFilePath);
+            var projectInfo = MiscellaneousFileUtilities.CreateMiscellaneousProjectInfoForDocument(
+                workspace, documentFilePath, sourceTextLoader, languageInformation, documentText.ChecksumAlgorithm, workspace.Services.SolutionServices, []);
+
+            // Ensure that syntax errors for `#:` directives are not reported on the primordial project.
+            Contract.ThrowIfNull(projectInfo.ParseOptions);
+            projectInfo = projectInfo.WithParseOptions(projectInfo.ParseOptions.WithFeatures([new("FileBasedProgram", "true")]));
+
+            workspace.OnProjectAdded(projectInfo);
+
+            if (languageInformation.LanguageName == "Razor")
+            {
+                var docId = projectInfo.AdditionalDocuments.Single().Id;
+                return workspace.CurrentSolution.GetRequiredAdditionalDocument(docId);
+            }
+
+            var id = projectInfo.Documents.Single().Id;
+            return workspace.CurrentSolution.GetRequiredDocument(id);
+        }
+    }
+
+    public async ValueTask TryRemoveMiscellaneousDocumentAsync(DocumentUri uri, bool removeFromMetadataWorkspace)
+    {
```

Comment:
duplicate code.  create helper.

Reply:
Which code is duplicated? Converting 'DocumentUri' to file path?

Reply:
Added helper.